Turn Template settings panel into a Template popover

noisysocks opened the pull request:
## What?
Rewrites the _Template_ panel in the sidebar to be a popover in the _Summary_ (formerly _Status & visibility_) panel.

## Why?
Part of https://github.com/WordPress/gutenberg/issues/39082.

## How?
- Removed `TemplatePanel` component from `@wordpress/edit-post`.
- Added `PostTemplate` component to `@wordpress/edit-post`.
  - This uses similar structure to existing `PostSchedule` and `PostVisibility` components.
  - All of the logic is the same as it was in `TemplatePanel`.
  - ~I've moved most of the logic inside a hook so that there is no code duplication between the label (`PostTemplate`) and the form (`PostTemplateForm`).~

## Testing Instructions
1. Activate a block theme.
2. Create or edit a post or a page.
3. Open the sidebar and change the post or page's template. Play around with editing templates and creating new templates.

## Screenshots or screencast 
https://user-images.githubusercontent.com/612155/175460161-1093568b-f2df-4591-a269-ff38bc9b9f74.mp4

noisysocks committed:
Rewrite Template panel as a popover

Mamaduka commented on packages/edit-post/src/components/sidebar/post-template/create-modal.js:
```diff
@@ -0,0 +1,150 @@
+const DEFAULT_TITLE = __( 'Custom Template' );
+
+export default function PostTemplateCreateModal( { onClose } ) {
+	const settings = useSelect(
+		( select ) => select( editorStore ).getEditorSettings(),
+		[]
+	);
```
Suggestion: I would recommend only returning the setting (`defaultBlockTemplate`) needed for this component to avoid re-rendering if any other editor's setting changes.

noisysocks replied:
I had to do this anyway cause splitting `useSelect` up was causing a bug due to calling `getEditedPostTemplate` on non block themes! 😀

Mamaduka commented on packages/edit-post/src/components/sidebar/post-template/create-modal.js:
```diff
@@ -0,0 +1,150 @@
+			] );
+
+		await __unstableCreateTemplate( {
+			slug: 'wp-custom-template-' + kebabCase( title || DEFAULT_TITLE ),
```
What do you think about using [`cleanForSlug`](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-url/#cleanforslug) here? Also, maybe we should get rid of the `wp-custom-template-` prefix. cc @mtias

noisysocks replied:
Both good ideas! Using `cleanForSlug` helps us eventually remove lodash, and the `wp-custom-template-`prefix doesn't seem necessary given that templates created in the site editor don't have a prefix. I double checked and if you create a template with a title that already exists the slug gets `-2` appended to it.

Mamaduka commented on packages/edit-post/src/components/sidebar/post-template/hook.js:
```diff
@@ -0,0 +1,110 @@
+		( settings.availableTemplates?.length ||
+			( settings.supportsTemplateMode && canUserCreateTemplate ) );
+
+	const isPostsPage = currentPostId === settings?.page_for_posts;
```
I think should be based on site settings and not the Editors - `const settings = getEntityRecord( 'root', 'site' );` .

noisysocks replied:
Is there a benefit to this approach? The nice thing about using editor settings is that we already have the settings object right there because we need it for `availableTemplates`. Worried about unnecessary REST API calls 😀

Mamaduka replied:
The `page_for_posts` isn't available in [Editor settings](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-admin/edit-form-blocks.php#L190-L210).

noisysocks replied:
Aha! Gotcha 😀

Mamaduka commented on packages/edit-post/src/components/sidebar/post-template/hook.js:
```diff
@@ -0,0 +1,110 @@
+		} );
+	}, [] );
+
+	const onEdit = useCallback( () => {
+		__unstableSwitchToTemplateMode();
+	}, [] );
```
Should we alias this directly, `onEdit: __unstableSwitchToTemplateMode`?

noisysocks replied:
You can't go `onClick={ __unstableSwitchToTemplateMode }` because then it thinks that the `event` object is the name of the template.

Mamaduka commented on packages/edit-post/src/components/sidebar/post-template/form.js:
```diff
@@ -0,0 +1,78 @@
+				}
+				onClose={ onClose }
+			/>
+			{ isPostsPage ? (
+				<Notice
+					className="edit-post-post-template__notice"
+					status="warning"
+					isDismissible={ false }
+				>
+					{ __( 'The posts page template cannot be changed.' ) }
+				</Notice>
```
This doesn't work because of the missing backport (https://github.com/WordPress/gutenberg/pull/38607#issuecomment-1138526934), but you can use the following snippet for testing.

```php
add_action( 'init', function() {
	register_setting(
		'reading',
		'page_for_posts',
		array(
			'show_in_rest' => true,
			'type'         => 'number',
			'description'  => __( 'The ID of the page that should display the latest posts', 'gutenberg' ),
		)
	);
} );
```

noisysocks replied:
Thanks! I was wondering what this was.

Mamaduka commented on packages/edit-post/src/components/sidebar/post-template/hook.js:
```diff
@@ -0,0 +1,110 @@
+		return getPostType( getCurrentPostType() )?.viewable ?? false;
+	}, [] );
+
+	const settings = useSelect(
+		( select ) => select( editorStore ).getEditorSettings(),
+		[]
+	);
```
Similar to my previous comment. I think we can just return `availableTemplates` and `supportsTemplateMode` here.

Mamaduka commented:
This change looks excellent. Fantastic work, @noisysocks!

I left a few suggestions, happy to do more testing later today.

Mamaduka commented on packages/edit-post/src/components/sidebar/post-template/form.js:
```diff
@@ -0,0 +1,78 @@
+
+	return (
+		<>
+			<InspectorPopoverHeader
```
I had no idea this component existed. It's really cool 🤩

noisysocks committed:
Don't call getEntityRecords and getEditedPostTemplate when isVisible is false

noisysocks committed:
Update e2e tests

noisysocks committed:
Update sidebar e2e test

talldan commented on packages/edit-post/src/components/sidebar/post-template/form.js:
```diff
@@ -0,0 +1,78 @@
+			<InspectorPopoverHeader
+				title={ __( 'Template' ) }
+				help={ __(
+					'Templates define the way your content is displayed when viewing your site.'
```
I think the first 'your' could be removed:
```suggestion
					'Templates define the way content is displayed when viewing your site.'
```

talldan commented on packages/edit-post/src/components/sidebar/post-template/hook.js:
```diff
@@ -0,0 +1,110 @@
+ */
+import { store as editPostStore } from '../../../store';
+
+export default function usePostTemplateForm() {
```
It doesn't return the form itself, so maybe it could be called `usePostTemplateFormProps`?

noisysocks replied:
Ended up removing this in last refactor!

talldan commented on packages/edit-post/src/components/sidebar/post-template/hook.js:
```diff
@@ -0,0 +1,110 @@
+	const canUserCreateTemplate = useSelect(
+		( select ) => select( coreStore ).canUser( 'create', 'templates' ),
+		[]
+	);
```
Might be overkill for this, but there's a `useResourcePermissions` hook that makes this easier - https://github.com/WordPress/gutenberg/pull/38785

That also tells you if the entity can be edited when an id is provided. I notice you're using the 'create' permission for that at the moment. I don't know if there's any difference.

noisysocks replied:
Good tip about `useResourcePermissions`. I think it's overkill here while we're only checking `create`.

I didn't want to change any of the `PostTemplatePanel` logic as part of this PR. We only check `create` in `trunk` so that's what I'm doing here. I'm not sure if we should check `edit`. Probably a good idea to investigate that separately.

talldan commented on packages/edit-post/src/components/sidebar/post-template/index.js:
```diff
@@ -0,0 +1,59 @@
+				position="bottom left"
+				contentClassName="edit-post-post-template__dialog"
+				renderToggle={ ( { isOpen, onToggle } ) => (
+					<Button
```
Should it have `aria-haspopup`? I'm not sure what the right value would be, possibly 'dialog'. The trouble with that is the popover itself doesn't have that role. 🤔 

Might be good to get some accessibility input on this.

https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-haspopup

noisysocks replied:
I'm not sure! 😅 If someone wants to weigh in and tell me what to do then I'm all ears.

All of the other popovers in the sidebar (post status, post date) currently work this way so I don't think it's a blocker. I'll be making more substantial changes to the panel as part of https://github.com/WordPress/gutenberg/issues/39082 so perhaps can do some thorough accessibility testing then. One thing I noticed while working on this PR is that the labels aren't great. For example the post status button's label is _Public_ which would sound strange out of context.

talldan commented on packages/edit-post/src/components/sidebar/post-template/style.scss:
```diff
@@ -0,0 +1,35 @@
+.edit-post-post-template__dialog {
+	z-index: z-index(".edit-post-post-template__dialog");
+
+	.components-popover__content {
```
I don't think it's good to override the `.components-popover__content` styles like this, but I see it's been done across the codebase, so too late now.

This classname can never be changed. 😬

mtias replied:
cc @ciampo @mirka for thoughts

mirka replied:
Any reason why we'd need to apply these styles directly on `.components-popover__content` though? It seems like the `width`/`padding` could just as easily be set on `<PostTemplateForm>` itself (or a div wrapper around it).

talldan replied:
Agreed, I think that would be a good alternative 👍

noisysocks replied:
Makes sense. I'm copying an existing pattern here but it's a small change so happy to change that for all of the popovers in this sidebar.

talldan commented on packages/edit-post/src/components/sidebar/post-template/style.scss:
```diff
@@ -0,0 +1,35 @@
+		}
+	}
+
+	.components-modal__header {
+		border-bottom: none;
+	}
+
+	.components-modal__content::before {
+		margin-bottom: $grid-unit-05;
+	}
```
Feels like this should be a prop on the `Modal` component to make it display this borderless variation. It'd be good to avoid overriding the component's style from outside.

It's brittle, because it can break when a component is refactored.

noisysocks replied:
Yeah totally agree. I'll do this separately as there's a few different parts of the codebase that we do this (welcome guide, list rename modal, start page options) and it warrants proper review.

noisysocks replied:
https://github.com/WordPress/gutenberg/pull/41998

talldan commented on packages/edit-post/src/components/sidebar/post-template/style.scss:
```diff
@@ -0,0 +1,35 @@
+		@include break-medium() {
+			width: $grid-unit * 40;
+		}
+	}
```
I reckon it'd be better to constrain the width of the `<form>` container rather than an individual element inside it.

talldan commented on packages/edit-post/src/components/sidebar/post-template/index.js:
```diff
@@ -0,0 +1,59 @@
+export default function PostTemplate() {
+	const anchorRef = useRef();
+
+	const postTemplateForm = usePostTemplateForm();
```
This hook does quite a lot, so I have a concern it might be expensive when a user is viewing the document sidebar. There's a strong chance they won't even interact with the template, so all that would be wasted computation. 

It looks like the only data that's needed in this situation is the template name. Maybe the rest can be deferred until the popover is opened.

noisysocks replied:
That's a really good point! I refactored this be more conventional (inline `useSelect` instead of `hook.js`) and in doing so made it so that `PostTemplate` loads the minimal amount of data necessary to render the label.

talldan reviewed (commented):
Great work. I didn't spot any issues when testing, so seems pretty close.

noisysocks committed:
Update iframed block e2e tests

noisysocks committed:
kebabCase -> cleanForSlug

noisysocks committed:
Restructure hooks so that we do minimal amount of work in PostTemplate label

noisysocks committed:
Remove 'your'

noisysocks commented:
Thanks for the reviews. I refactored the code a little bit so that we're not performing unnecessary data fetching. I'll work on addressing the components / styling feedback tomorrow.

javierarce commented:
This looks fine! I have just two comments (maybe out of the scope for this PR):

- Long template names get too closer to the label and overflow the row. Could we consider truncating the string and adding a tooltip (like we do with the Publish field)?
- We load the full Template field at once after we finish loading the list of templates. This introduces some movement to the other sidebar elements. Would it be possible to show the Template row while we are fetching the list of templates? 
 
Regarding the icon we use to add templates, I've created [a proposal for a new one here](https://www.figma.com/file/e4tLacmlPuZV47l7901FEs/WordPress-Design-Library?node-id=37257%3A25895):
 
![image](https://user-images.githubusercontent.com/4933/175903391-e2d0523f-e134-4cdf-b8e0-19e2e9bdb980.png)

@WordPress/gutenberg-design

noisysocks committed:
Use site settings for page_for_posts

noisysocks committed:
Don't style .components-popover__content directly

noisysocks committed:
Constrain form width, not text field

noisysocks committed:
Fix empty availableTemplates check

noisysocks committed:
Truncate Template button text

noisysocks committed:
Use new add template icon

noisysocks commented:
> * Long template names get too closer to the label and overflow the row. Could we consider truncating the string and adding a tooltip (like we do with the Publish field)?

Good flag, I've made the text truncate.

> * We load the full Template field at once after we finish loading the list of templates. This introduces some movement to the other sidebar elements. Would it be possible to show the Template row while we are fetching the list of templates?

I've fixed a bug which was making this pop-in slower than it should be, so this looks better now.

I can't _completely_ remove the pop-in. There seems to be a bug with how caching works in `@wordpres/core-data` which means that it always takes at least a frame for cached data to be returned from `getEntityRecord` because of locking. I'll try to investigate this later, maybe with @adamziel's help.

We could show an empty row or a spinner while we're waiting for the cached data but then this would look weird for users that do not have the necessary permissions to change the template.

> Regarding the icon we use to add templates, I've created [a proposal for a new one here](https://www.figma.com/file/e4tLacmlPuZV47l7901FEs/WordPress-Design-Library?node-id=37257%3A25895):

Oh yep, thanks! Switched to that new icon.

noisysocks committed:
Replace 'Status & visibility' with 'Summary' in docs

noisysocks commented:
OK think I've responded to everything and that this is ready for re-review!

Mamaduka commented on packages/edit-post/src/components/sidebar/post-template/create-modal.js:
```diff
@@ -0,0 +1,145 @@
+const DEFAULT_TITLE = __( 'Custom Template' );
+
+export default function PostTemplateCreateModal( { onClose } ) {
+	const settings = useSelect(
+		( select ) => select( editorStore ).getEditorSettings(),
+		[]
+	);
```
```suggestion
	const defaultBlockTemplate = useSelect(
		( select ) =>
			select( editorStore ).getEditorSettings()?.defaultBlockTemplate,
		[]
	);
```

Nitpick (non-blocking): Since we only use `defaultBlockTemplate` in the component, let's optimize the selector.

Mamaduka reviewed (approved):
This tests well for me. Thanks for fantastic work, @noisysocks 🦸

I left a small nitpick comment, but feel free to ignore it.

@javierarce, the "page for posts" notice might not fit well in your design, but happy to polish this as a follow-up.

## Screenshot of notice
![CleanShot 2022-06-28 at 11 04 26](https://user-images.githubusercontent.com/240569/176116422-feaa6c67-6899-4936-96a6-9b7ea4753cc0.png)

noisysocks committed:
Return only defaultBlockTemplate from useSelect

jameskoster commented:
I think the changes here are generally an improvement. My one concern is that we may go in a completely different direction when we address https://github.com/WordPress/gutenberg/issues/41258. Would it be harmful to repeatedly change the template UX?

javierarce commented:
If there's a strong intention to go in that other direction and a timeframe for the change, I think it would probably make sense to reconsider this PR (but that's my personal opinion).

Otherwise, I think the changes here (which are part of the changes required for #39082 and attempt to unify the way users manage the settings) make sense for the near future.

jameskoster commented:
Yeah it's a tough one... I don't think it's 100% clear which direction we'll head in just yet. 

Maybe since the Template panel _is not currently a part_ of the Summary section we should consider it out of scope in the context of https://github.com/WordPress/gutenberg/issues/39082? Especially since this PR doesn't address the underlying UX issues of that panel (making template selection a more visual experience).

noisysocks commented:
Ooh. If I had seen https://github.com/WordPress/gutenberg/issues/41258 before starting on this then, yeah, it might have been worth holding off. But, I didn't! 😅 Since the work has been done and is ready to go, I think we may as well ship this as a small improvement on the path towards a more substantial improvement. I don't see a downside as this component is not a public API so it's quite easy to change or remove in the future.

noisysocks commented:
I'm going to go ahead and merge. We can continue to iterate as always. Thanks everyone for the really great feedback!

paaljoachim commented:
Before:
![Screen Shot 2022-07-05 at 00 28 03](https://user-images.githubusercontent.com/5323259/177463050-5c9a47b9-7a3f-4197-a983-358d3cdda4f0.png)

After:
![Screen Shot 2022-07-05 at 00 28 57](https://user-images.githubusercontent.com/5323259/177463077-c32e2ca9-d437-4b6d-8fb3-744477ec6b5f.png)

To me the result looks like a nice upgrade to what was before.

This does feel like a part of the Summary section though, so that this PR should be linked in with the Summary section issue. Looking at the UX used in this PR for the Template to see if it is natural to bring it onward to other areas inside the Summary section.

noisysocks commented:
> This does feel like a part of the Summary section though, so that this PR should be linked in with the Summary section issue. 

Do you mean https://github.com/WordPress/gutenberg/issues/39082? It's linked.
